# EDA 1: Introducción

In [83]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [8]:
df = pd.read_csv("datos/EDA", index_col=0)
df.head()

case_number    year        type    country        area  \
0    2018.06.25  2018.0     Boating        USA  California   
1  2018.06.03.a  2018.0  Unprovoked     BRAZIL  Pernambuco   
2  2018.05.26.b  2018.0  Unprovoked        USA     Florida   
3    2018.05.24  2018.0    Provoked  AUSTRALIA  Queensland   
4  2018.05.13.a  2018.0     Invalid    ENGLAND    Cornwall   

                       location        activity                    name age  \
0   Oceanside, San Diego County        Paddling             Julie Wolfe  57   
1         Piedade Beach, Recife        Swimming  Jose Ernesto da Silva   18   
2  Cocoa Beach, Brevard  County         Walking               Cody High  15   
3               Cairns Aquarium  Feeding sharks                    male  32   
4                Off Land's End         Fishing            Max Berryman  21   

             time          species_  \
0           18h00       White shark   
1  Late afternoon       Tiger shark   
2           17h00    Bull shark, 6'   
3         Morning   Grey reef shark   
4           08h15  Invalid incident   

                                                href      siglo  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
3  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
4  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   

                                              injury month fatal sex  
0  No injury to occupant, outrigger canoe and pad...   Jun     N   F  
1                                              FATAL   Jun     Y   M  
2                              Lower left leg bitten   May     N   M  
3  Minor bite to hand by captive shark. PROVOKED ...   May     N   M  
4  Injured by teeth of a dead porbeagle shark he ...   May     N   M

## Hipótesis

Las hipótesis que os planteamos son:

- ¿Es Australia el sitio más peligroso y letal para estar relajada en la playa?

- ¿Hay diferencias entre los países en los ataques a hombres y mujeres?

- ¿Cuáles son las edades que más sufren ataques?

- Independientemente de la edad, ¿sufren los hombres más ataques que las mujeres?

- ¿En qué mes ocurren más ataques?

- ¿Cuál es la relación entre la especie y el tipo de ataque (si es fatal o no)?

- ¿Cómo han evolucionado los ataques a lo largo del tiempo?




## Objetivos

Los objetivos de hoy son:

#### 1. Haced una exploración inicial del dataframe:
  
- ¿Cuál es el número de filas y columnas?

In [9]:
df.shape

(1672, 17)

- ¿Cuál es el porcentaje de valores nulos en cada columna? Presenta los resultados en un dataframe, donde las columnas sean el nombre de las variables y el porcentaje de nulos.

In [132]:
nulos = pd.DataFrame(round(df.isnull().sum() * 100 / df.shape[0], 3)).T
nulos

case_number  year  type  country   area  location  activity   name    age  \
0          0.0   0.0   0.0    0.598  2.751      3.05     1.854  0.538  9.211   

    time  species_  href  siglo  injury   month  fatal    sex  
0  9.569     7.536   0.0    0.0   0.478  10.825  5.921  0.837

- ¿Qué tipos de datos hay en cada columna?

In [20]:
df.dtypes

case_number     object
year           float64
type            object
country         object
area            object
location        object
activity        object
name            object
age             object
time            object
species_        object
href            object
siglo           object
injury          object
month           object
fatal           object
sex             object
dtype: object

- Extraed la información general del dataframe.

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1672 entries, 0 to 1671
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   case_number  1672 non-null   object 
 1   year         1672 non-null   float64
 2   type         1672 non-null   object 
 3   country      1662 non-null   object 
 4   area         1626 non-null   object 
 5   location     1621 non-null   object 
 6   activity     1641 non-null   object 
 7   name         1663 non-null   object 
 8   age          1518 non-null   object 
 9   time         1512 non-null   object 
 10  species_     1546 non-null   object 
 11  href         1672 non-null   object 
 12  siglo        1672 non-null   object 
 13  injury       1664 non-null   object 
 14  month        1491 non-null   object 
 15  fatal        1573 non-null   object 
 16  sex          1658 non-null   object 
dtypes: float64(1), object(16)
memory usage: 235.1+ KB


- Cread dos dataframes nuevos, uno que incluya solo las variables categóricas y otro que incluya solo las variables numéricas.

In [128]:
df_numerica = df.select_dtypes(include = "float64")
df_numerica.head()

year
0  2018.0
1  2018.0
2  2018.0
3  2018.0
4  2018.0

In [133]:
df_categ = df.select_dtypes(include = "object")
df_categ.head()

case_number        type    country        area  \
0    2018.06.25     Boating        USA  California   
1  2018.06.03.a  Unprovoked     BRAZIL  Pernambuco   
2  2018.05.26.b  Unprovoked        USA     Florida   
3    2018.05.24    Provoked  AUSTRALIA  Queensland   
4  2018.05.13.a     Invalid    ENGLAND    Cornwall   

                       location        activity                    name age  \
0   Oceanside, San Diego County        Paddling             Julie Wolfe  57   
1         Piedade Beach, Recife        Swimming  Jose Ernesto da Silva   18   
2  Cocoa Beach, Brevard  County         Walking               Cody High  15   
3               Cairns Aquarium  Feeding sharks                    male  32   
4                Off Land's End         Fishing            Max Berryman  21   

             time          species_  \
0           18h00       White shark   
1  Late afternoon       Tiger shark   
2           17h00    Bull shark, 6'   
3         Morning   Grey reef shark   
4           08h15  Invalid incident   

                                                href      siglo  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
3  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   
4  http://sharkattackfile.net/spreadsheets/pdf_di...  siglo-XXI   

                                              injury month fatal sex  
0  No injury to occupant, outrigger canoe and pad...   Jun     N   F  
1                                              FATAL   Jun     Y   M  
2                              Lower left leg bitten   May     N   M  
3  Minor bite to hand by captive shark. PROVOKED ...   May     N   M  
4  Injured by teeth of a dead porbeagle shark he ...   May     N   M

- ¿Cuáles son los principales estádisticos de los dataframes?

In [34]:
df.describe(include= "all").T

count unique                                                top  \
case_number    1672   1672                                         2018.06.25   
year         1672.0    NaN                                                NaN   
type           1672      6                                         Unprovoked   
country        1662     95                                                USA   
area           1626    236                                            Florida   
location       1621   1310                   New Smyrna Beach, Volusia County   
activity       1641    404                                            Surfing   
name           1663   1536                                               male   
age            1518     94                                                 19   
time           1512    248                                              11h00   
species_       1546    856                                        White shark   
href           1672   1672  http://sharkattackfile.net/spreadsheets/pdf_di...   
siglo          1672      2                                           siglo-XX   
injury         1664   1173                                              FATAL   
month          1491     12                                                Aug   
fatal          1573      2                                                  N   
sex            1658      2                                                  M   

             freq         mean         std  min     25%     50%     75%  \
case_number     1          NaN         NaN  NaN     NaN     NaN     NaN   
year          NaN  1832.688397  528.969901  0.0  1964.0  1997.0  2008.0   
type         1432          NaN         NaN  NaN     NaN     NaN     NaN   
country       741          NaN         NaN  NaN     NaN     NaN     NaN   
area          346          NaN         NaN  NaN     NaN     NaN     NaN   
location       73          NaN         NaN  NaN     NaN     NaN     NaN   
activity      426          NaN         NaN  NaN     NaN     NaN     NaN   
name           97          NaN         NaN  NaN     NaN     NaN     NaN   
age            70          NaN         NaN  NaN     NaN     NaN     NaN   
time           74          NaN         NaN  NaN     NaN     NaN     NaN   
species_       76          NaN         NaN  NaN     NaN     NaN     NaN   
href            1          NaN         NaN  NaN     NaN     NaN     NaN   
siglo         887          NaN         NaN  NaN     NaN     NaN     NaN   
injury        154          NaN         NaN  NaN     NaN     NaN     NaN   
month         156          NaN         NaN  NaN     NaN     NaN     NaN   
fatal        1256          NaN         NaN  NaN     NaN     NaN     NaN   
sex          1435          NaN         NaN  NaN     NaN     NaN     NaN   

                max  
case_number     NaN  
year         2018.0  
type            NaN  
country         NaN  
area            NaN  
location        NaN  
activity        NaN  
name            NaN  
age             NaN  
time            NaN  
species_        NaN  
href            NaN  
siglo           NaN  
injury          NaN  
month           NaN  
fatal           NaN  
sex             NaN

- Extraed los valores únicos su frecuencia. BONUS presenta los resultados en un dataframe. 

In [130]:
# Creamos dos listas vacías, para poder crear un diccionario con los nombres de cada columna, y los valores únicos de cada una:

lista = []
valores = []

for col in df.columns:
    lista.append(col)
    valores.append(df[col].unique())
dicc = {"variable":lista, "n_uniques":valores}

In [131]:
df_vu = pd.DataFrame.from_dict(dicc)
df_vu

variable                                          n_uniques
0   case_number  [2018.06.25, 2018.06.03.a, 2018.05.26.b, 2018....
1          year  [2018.0, 2017.0, 2016.0, 2015.0, 2014.0, 2013....
2          type  [Boating, Unprovoked, Provoked, Invalid, Sea D...
3       country  [USA, BRAZIL, AUSTRALIA, ENGLAND, COSTA RICA, ...
4          area  [California, Pernambuco, Florida, Queensland, ...
5      location  [Oceanside, San Diego County, Piedade Beach, R...
6      activity  [Paddling, Swimming, Walking, Feeding sharks, ...
7          name  [Julie Wolfe, Jose Ernesto da Silva , Cody Hig...
8           age  [57, 18, 15, 32, 21, 30, 60, 33, 19, 25, 10, 6...
9          time  [18h00, Late afternoon, 17h00, Morning, 08h15,...
10     species_  [White shark, Tiger shark, Bull shark, 6', Gre...
11         href  [http://sharkattackfile.net/spreadsheets/pdf_d...
12        siglo                              [siglo-XXI, siglo-XX]
13       injury  [No injury to occupant, outrigger canoe and pa...
14        month  [Jun, May, Apr, Mar, Feb, Jan, Dec, Nov, Oct, ...
15        fatal                                        [N, Y, nan]
16          sex                                        [F, M, nan]

## Conclusiones

De qué nos sirven estos números y dataframe? Extraed conclusiones de ellos.

Por ejemplo:

##### - ¿Hay alguna columna que no tenga sentido su tipo de datos?

Si, hay dos columnas: 

- "year" aparece como float, por lo que los años aparecen con decimales (2018.0). Sería más lógico que fuera datetime o int.

- "age" aparece como object, y lo lógico sería que fuera int.

##### - ¿Hay alguna columna que no tenga sentido su nombre?

Sí, la columna "href". Contiene un enlaces a un pdf, por lo que entendemos que su signficado es "reference". Pero para una persona que no tiene acceso a la documentación que explique el nombre de las columnas y su contenido, puede ser confuso.

##### - ¿Hay alguna columna que no tenga sentido su contenido?

La columna "time" se clasifica por horas o momentos del día, lo que la hace ilegible y difícil de analizar, dado que su clasificación no está homogeneizada.

##### - ¿Hay alguna columna que tenga demasiadas categorías o valores que nos pueda hacer difícil trabajar con ella?

La columna "injury" tiene 1173 valores únicos, lo que para un dataset de 1600 registros, es una barbaridad. Dificulta la extracción de frecuencias y un análisis riguroso.